File: lib/services/cita_service.dart
import '../config/constants.dart';
import '../database/database_helper.dart';
import '../models/cita.dart';
import '../models/ingreso.dart';
import 'cliente_service.dart';
import 'finanzas_service.dart';

class CitaService {
  final DatabaseHelper _db = DatabaseHelper();
  final FinanzasService _finanzas = FinanzasService();
  final ClienteService _clientes = ClienteService();

  // Crear cita
  Future<int> crearCita(Cita cita) async {
    final citaConFecha = cita.copyWith(
      fechaCreacion: cita.fechaCreacion ?? DateTime.now(),
    );
    final id = await _db.insertCita(citaConFecha.toMap());
    await sincronizarIngreso(citaConFecha.copyWith(id: id));
    return id;
  }

  /// Mantiene el ingreso en sincronía con el estado de la cita:
  /// - Si la cita queda *completada* y tiene monto, registra (o actualiza) un
  ///   ingreso enlazado a la cita.
  /// - Si la cita deja de estar completada, elimina el ingreso asociado.
  /// Es idempotente: no crea ingresos duplicados para la misma cita.
  Future<void> sincronizarIngreso(Cita cita) async {
    if (cita.id == null) {
      return;
    }
    final existentes = await _finanzas.obtenerIngresosPorCita(cita.id!);
    final debeTenerIngreso =
        cita.estado == EstadoCita.completada && (cita.monto ?? 0) > 0;

    // Al completar una cita, registra la visita en la ficha del cliente
    // (sirva de donde sirva la acción: formulario o agenda).
    if (cita.estado == EstadoCita.completada) {
      await _clientes.marcarUltimaVisita(cita.clienteId, cita.fechaHora);
    }

    if (!debeTenerIngreso) {
      if (existentes.isNotEmpty) {
        await _finanzas.eliminarIngresosPorCita(cita.id!);
      }
      return;
    }

    // Debe existir un ingreso que refleje el monto y la fecha actuales.
    final yaCorrecto = existentes.length == 1 &&
        existentes.first.monto == cita.monto &&
        existentes.first.fecha == cita.fechaHora;
    if (yaCorrecto) {
      return;
    }
    if (existentes.isNotEmpty) {
      await _finanzas.eliminarIngresosPorCita(cita.id!);
    }
    await _finanzas.registrarIngreso(
      Ingreso(
        citaId: cita.id,
        monto: cita.monto!,
        metodo: AppConstants.metodosPago.first,
        fecha: cita.fechaHora,
        notas: 'Generado automáticamente por cita completada',
      ),
    );
  }

  // Obtener todas las citas
  Future<List<Cita>> obtenerTodas() async {
    final mapList = await _db.getAllCitas();
    return mapList.map((map) => Cita.fromMap(map)).toList();
  }

  // Obtener citas por fecha
  Future<List<Cita>> obtenerPorFecha(DateTime fecha) async {
    final mapList = await _db.getCitasByFecha(fecha);
    return mapList.map((map) => Cita.fromMap(map)).toList();
  }

  // Obtener citas por cliente
  Future<List<Cita>> obtenerPorCliente(int clienteId) async {
    final mapList = await _db.getCitasByCliente(clienteId);
    return mapList.map((map) => Cita.fromMap(map)).toList();
  }

  // Obtener citas de hoy
  Future<List<Cita>> obtenerHoy() async {
    return await obtenerPorFecha(DateTime.now());
  }

  // Obtener citas de la semana
  Future<List<Cita>> obtenerSemana() async {
    final hoy = DateTime.now();
    final citas = await obtenerTodas();
    return citas
        .where((c) =>
            c.fechaHora.isAfter(hoy.subtract(const Duration(days: 1))) &&
            c.fechaHora.isBefore(hoy.add(const Duration(days: 8))))
        .toList();
  }

  // Actualizar cita
  Future<int> actualizar(Cita cita) async {
    final resultado = await _db.updateCita(cita.toMap());
    await sincronizarIngreso(cita);
    return resultado;
  }

  // Eliminar cita (elimina también el ingreso asociado, si lo hubiera)
  Future<int> eliminar(int id) async {
    await _finanzas.eliminarIngresosPorCita(id);
    return await _db.deleteCita(id);
  }

  // Eliminar todas las citas de un cliente (con sus ingresos). Se usa al
  // borrar un cliente que no tiene citas completadas.
  Future<void> eliminarPorCliente(int clienteId) async {
    final citas = await obtenerPorCliente(clienteId);
    for (final cita in citas) {
      if (cita.id != null) {
        await eliminar(cita.id!);
      }
    }
  }

  // Cambiar estado de cita
  Future<int> cambiarEstado(int id, EstadoCita nuevoEstado) async {
    final citas = await obtenerTodas();
    final cita = citas.firstWhere((c) => c.id == id);
    final citaActualizada = cita.copyWith(estado: nuevoEstado);
    return await actualizar(citaActualizada);
  }

  // Obtener citas pendientes
  Future<List<Cita>> obtenerPendientes() async {
    final citas = await obtenerTodas();
    return citas
        .where((c) => c.estado == EstadoCita.pendiente)
        .toList();
  }

  // Obtener citas confirmadas
  Future<List<Cita>> obtenerConfirmadas() async {
    final citas = await obtenerTodas();
    return citas
        .where((c) => c.estado == EstadoCita.confirmada)
        .toList();
  }

  // Obtener citas completadas
  Future<List<Cita>> obtenerCompletadas() async {
    final citas = await obtenerTodas();
    return citas
        .where((c) => c.estado == EstadoCita.completada)
        .toList();
  }

  // Obtener total de citas hoy
  Future<int> totalHoy() async {
    final hoy = await obtenerHoy();
    return hoy.length;
  }

  // Obtener total de ingresos hoy
  Future<double> ingresosHoy() async {
    final hoy = await obtenerHoy();
    return hoy
        .where((c) => c.estado == EstadoCita.completada)
        .fold<double>(0, (sum, c) => sum + (c.monto ?? 0));
  }

  // Verificar disponibilidad en horario
  Future<bool> estaDisponible(DateTime fechaHora, int duracionMinutos) async {
    final citasDelDia = await obtenerPorFecha(fechaHora);
    final citasConfirmadas = citasDelDia
        .where((c) => c.estado == EstadoCita.confirmada)
        .toList();

    for (var cita in citasConfirmadas) {
      final inicio = cita.fechaHora;
      final fin = inicio.add(Duration(minutes: cita.duracionMinutos));

      if (fechaHora.isBefore(fin) &&
          fechaHora.add(Duration(minutes: duracionMinutos)).isAfter(inicio)) {
        return false;
      }
    }
    return true;
  }

  // Exportar citas a mapas
  Future<List<Map<String, dynamic>>> exportarTodas() async {
    final citas = await obtenerTodas();
    return citas.map((c) => c.toMap()).toList();
  }
}

File: qt/src/services/CitaService.cpp
#include "services/CitaService.h"

#include "db/Database.h"
#include "services/ClienteService.h"
#include "services/FinanzasService.h"

#include <QDate>
#include <QDateTime>
#include <QSqlQuery>
#include <QVariant>

const QString CitaService::kSelect = QStringLiteral(
    "SELECT c.id, c.cliente_id AS clienteId, c.servicio_id AS servicioId, "
    "c.fecha_hora AS fechaHora, c.duracion_minutos AS duracionMinutos, "
    "c.estado, c.monto, c.notas, c.fecha_creacion AS fechaCreacion, "
    "cl.nombre AS nombreCliente, s.nombre AS nombreServicio "
    "FROM citas c "
    "LEFT JOIN clientes cl ON c.cliente_id = cl.id "
    "LEFT JOIN servicios s ON c.servicio_id = s.id ");

CitaService::CitaService(FinanzasService *finanzas, ClienteService *clientes,
                         QObject *parent)
    : QObject(parent), m_finanzas(finanzas), m_clientes(clientes)
{
}

QVariantList CitaService::obtenerTodas() const
{
    QSqlQuery q = Database::instance().exec(kSelect + QStringLiteral("ORDER BY c.fecha_hora DESC"));
    return Database::rows(q);
}

QVariantList CitaService::obtenerPorFecha(const QString &fechaIso) const
{
    const QString dia = fechaIso.left(10); // yyyy-MM-dd
    QSqlQuery q = Database::instance().exec(
        kSelect + QStringLiteral("WHERE DATE(c.fecha_hora) = ? ORDER BY c.fecha_hora"),
        {dia});
    return Database::rows(q);
}

QVariantList CitaService::obtenerPorCliente(int clienteId) const
{
    QSqlQuery q = Database::instance().exec(
        kSelect + QStringLiteral("WHERE c.cliente_id = ? ORDER BY c.fecha_hora DESC"),
        {clienteId});
    return Database::rows(q);
}

QVariantList CitaService::activasPorFecha(const QString &fechaIso) const
{
    const QString dia = fechaIso.left(10);
    QSqlQuery q = Database::instance().exec(
        kSelect + QStringLiteral("WHERE DATE(c.fecha_hora) = ? AND c.estado IN "
                                 "('pendiente','confirmada') ORDER BY c.fecha_hora"),
        {dia});
    return Database::rows(q);
}

QVariantList CitaService::historial() const
{
    QSqlQuery q = Database::instance().exec(
        kSelect + QStringLiteral("WHERE c.estado IN ('completada','cancelada') "
                                 "ORDER BY c.fecha_hora DESC"));
    return Database::rows(q);
}

int CitaService::crear(const QVariantMap &datos)
{
    QString fechaCreacion = datos.value(QStringLiteral("fechaCreacion")).toString();
    if (fechaCreacion.isEmpty())
        fechaCreacion = QDateTime::currentDateTime().toString(Qt::ISODate);
    const QString estado = datos.value(QStringLiteral("estado"),
                                       QStringLiteral("pendiente")).toString();

    QSqlQuery q = Database::instance().exec(
        QStringLiteral("INSERT INTO citas (cliente_id, servicio_id, fecha_hora, "
                       "duracion_minutos, estado, monto, notas, fecha_creacion) "
                       "VALUES (?, ?, ?, ?, ?, ?, ?, ?)"),
        {datos.value(QStringLiteral("clienteId")),
         datos.value(QStringLiteral("servicioId")),
         datos.value(QStringLiteral("fechaHora")),
         datos.value(QStringLiteral("duracionMinutos")),
         estado,
         datos.value(QStringLiteral("monto")),
         datos.value(QStringLiteral("notas")),
         fechaCreacion});

    const QVariant nuevoId = q.lastInsertId();
    if (!nuevoId.isValid())
        return -1;
    const int id = nuevoId.toInt();
    sincronizarIngreso(id, datos.value(QStringLiteral("clienteId")).toInt(), estado,
                       datos.value(QStringLiteral("monto")),
                       datos.value(QStringLiteral("fechaHora")).toString());
    emit cambiado();
    return id;
}

bool CitaService::actualizar(const QVariantMap &datos)
{
    if (!datos.contains(QStringLiteral("id")))
        return false;
    const int id = datos.value(QStringLiteral("id")).toInt();
    const QString estado = datos.value(QStringLiteral("estado"),
                                       QStringLiteral("pendiente")).toString();

    Database::instance().exec(
        QStringLiteral("UPDATE citas SET cliente_id = ?, servicio_id = ?, fecha_hora = ?, "
                       "duracion_minutos = ?, estado = ?, monto = ?, notas = ? WHERE id = ?"),
        {datos.value(QStringLiteral("clienteId")),
         datos.value(QStringLiteral("servicioId")),
         datos.value(QStringLiteral("fechaHora")),
         datos.value(QStringLiteral("duracionMinutos")),
         estado,
         datos.value(QStringLiteral("monto")),
         datos.value(QStringLiteral("notas")),
         id});

    sincronizarIngreso(id, datos.value(QStringLiteral("clienteId")).toInt(), estado,
                       datos.value(QStringLiteral("monto")),
                       datos.value(QStringLiteral("fechaHora")).toString());
    emit cambiado();
    return true;
}

bool CitaService::eliminar(int id)
{
    // Una cita completada ya generó su ingreso: borrarla se llevaría el
    // historial de facturación por delante. Se protege aquí, no solo en la UI.
    QSqlQuery info = Database::instance().exec(
        QStringLiteral("SELECT estado FROM citas WHERE id = ?"), {id});
    if (info.next() && info.value(0).toString() == QStringLiteral("completada"))
        return false;

    // Elimina también el ingreso asociado, si lo hubiera.
    m_finanzas->eliminarIngresosPorCita(id);
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("DELETE FROM citas WHERE id = ?"), {id});
    const bool ok = q.numRowsAffected() > 0;
    if (ok)
        emit cambiado();
    return ok;
}

bool CitaService::cambiarEstado(int id, const QString &estado)
{
    QSqlQuery info = Database::instance().exec(
        QStringLiteral("SELECT cliente_id, monto, fecha_hora FROM citas WHERE id = ?"),
        {id});
    if (!info.next())
        return false;
    const int clienteId = info.value(0).toInt();
    const QVariant monto = info.value(1);
    const QString fechaHora = info.value(2).toString();

    Database::instance().exec(
        QStringLiteral("UPDATE citas SET estado = ? WHERE id = ?"), {estado, id});
    sincronizarIngreso(id, clienteId, estado, monto, fechaHora);
    emit cambiado();
    return true;
}

// Mantiene el ingreso en sincronía con el estado de la cita. Idempotente.
void CitaService::sincronizarIngreso(int id, int clienteId, const QString &estado,
                                     const QVariant &monto, const QString &fechaHora)
{
    if (id <= 0)
        return;
    const QVariantList existentes = m_finanzas->obtenerIngresosPorCita(id);
    const double montoVal = monto.toDouble();
    const bool debeTenerIngreso =
        estado == QStringLiteral("completada") && montoVal > 0;

    // Al completar, registra la visita en la ficha del cliente.
    if (estado == QStringLiteral("completada"))
        m_clientes->marcarUltimaVisita(clienteId, fechaHora);

    if (!debeTenerIngreso) {
        if (!existentes.isEmpty())
            m_finanzas->eliminarIngresosPorCita(id);
        return;
    }

    // Debe existir un único ingreso que refleje el monto y la fecha actuales.
    bool yaCorrecto = false;
    if (existentes.size() == 1) {
        const QVariantMap ing = existentes.first().toMap();
        yaCorrecto = ing.value(QStringLiteral("monto")).toDouble() == montoVal &&
                     ing.value(QStringLiteral("fecha")).toString() == fechaHora;
    }
    if (yaCorrecto)
        return;

    if (!existentes.isEmpty())
        m_finanzas->eliminarIngresosPorCita(id);

    QVariantMap nuevo;
    nuevo.insert(QStringLiteral("citaId"), id);
    nuevo.insert(QStringLiteral("monto"), montoVal);
    nuevo.insert(QStringLiteral("metodo"), QStringLiteral("Efectivo")); // metodosPago.first
    nuevo.insert(QStringLiteral("fecha"), fechaHora);
    nuevo.insert(QStringLiteral("notas"),
                 QStringLiteral("Generado automáticamente por cita completada"));
    m_finanzas->registrarIngreso(nuevo);
}

int CitaService::totalHoy() const
{
    const QString hoy = QDate::currentDate().toString(QStringLiteral("yyyy-MM-dd"));
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("SELECT COUNT(*) FROM citas WHERE DATE(fecha_hora) = ?"), {hoy});
    return q.next() ? q.value(0).toInt() : 0;
}

double CitaService::ingresosHoy() const
{
    const QString hoy = QDate::currentDate().toString(QStringLiteral("yyyy-MM-dd"));
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("SELECT COALESCE(SUM(monto), 0) FROM citas "
                       "WHERE estado = 'completada' AND DATE(fecha_hora) = ?"),
        {hoy});
    return q.next() ? q.value(0).toDouble() : 0.0;
}

bool CitaService::estaDisponible(const QString &fechaHoraIso, int duracionMinutos) const
{
    const QDateTime inicioNuevo = QDateTime::fromString(fechaHoraIso, Qt::ISODate);
    if (!inicioNuevo.isValid())
        return true;
    const QDateTime finNuevo = inicioNuevo.addSecs(duracionMinutos * 60);

    const QString dia = fechaHoraIso.left(10);
    QSqlQuery q = Database::instance().exec(
        QStringLiteral("SELECT fecha_hora, duracion_minutos FROM citas "
                       "WHERE DATE(fecha_hora) = ? AND estado = 'confirmada'"),
        {dia});
    while (q.next()) {
        const QDateTime inicio = QDateTime::fromString(q.value(0).toString(), Qt::ISODate);
        const QDateTime fin = inicio.addSecs(q.value(1).toInt() * 60);
        if (inicioNuevo < fin && finNuevo > inicio)
            return false;
    }
    return true;
}
